src/v/net/batched_output_stream.cc

```diff
@@ -37,24 +37,27 @@ already_closed_error(ss::scattered_message<char>& msg) {
 }
 
 ss::future<bool> batched_output_stream::write(ss::scattered_message<char> msg) {
```

Comment:
Sorry, I don't fully understand the original bug. Is the scenario that `stop()` is called on the stream (but the stream is not destroyed), while there are still writes in flight?

Reply:
Now that you asked i think the problem is in different place, it is in `http::client`

Reply:
The following situations seems possible: 
calling `client::send()` and without waiting for result calling `client::stop()` this calls the underlying `base_transport::stop()`. The `base_transport::stop()` in the end destroys the `_out` stream. It waits for `_dispatch_gate` to be closed but `http::client` never actually holds that gate

Reply:
Just to be clear, we are talking about `http::client` here, right?

> The following situations seems possible:
calling client::send() and without waiting for result calling client::stop() this calls the underlying base_transport::stop(). The base_transport::stop() in the end destroys the _out stream. It waits for _dispatch_gate to be closed but http::client never actually holds that gate

Hmm but `buffered_output_stream::_write_sem` will be held by `send -> bos::write()`, so when `client::stop` calls `bos::stop` `bos::stop` which tries to grab that same semaphore. So by the time `bos::stop()` future resolves any earlier calls to `write` should have resolved (since they wait on the semaphore in order).